File: ch04/spring-boot-actuator-custom-metrics/spring-boot-actuator-custom-metrics-final/src/main/java/com/manning/sbip/ch04/service/DefaultCourseService.java
package com.manning.sbip.ch04.service;

import com.manning.sbip.ch04.model.Course;
import com.manning.sbip.ch04.repository.CourseRepository;
import io.micrometer.core.instrument.Counter;
import io.micrometer.core.instrument.DistributionSummary;
import io.micrometer.core.instrument.Timer;
import lombok.SneakyThrows;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.Optional;
import java.util.concurrent.Callable;

@Service
public class DefaultCourseService implements CourseService {

    private final CourseRepository courseRepository;
    private final Counter createCourseCounter;
    private final Timer createCoursesTimer;
    private final DistributionSummary distributionSummary;

    @Autowired
    public DefaultCourseService(CourseRepository courseRepository,
                                Counter createCourseCounter,
                                Timer createCoursesTimer,
                                DistributionSummary distributionSummary) {
        this.courseRepository = courseRepository;
        this.createCourseCounter = createCourseCounter;
        this.createCoursesTimer = createCoursesTimer;
        this.distributionSummary = distributionSummary;
    }

    @SneakyThrows
    public Course createCourse(Course course) {

        /**
         * Below statement is used to count the number of courses created
         */
        createCourseCounter.increment();
        //return courseRepository.save(course);

        distributionSummary.record(course.getRating());

        /**
         * Below statement is used to capture the time taken to save
         * a course into the database using the Timer
         */
        return createCoursesTimer.recordCallable(() -> courseRepository.save(course));

    }

    public Optional<Course> findCourseById(Long courseId) {
        return courseRepository.findById(courseId);
    }

    public Iterable<Course> findAllCourses() {
        return courseRepository.findAll();
    }

    public Course updateCourse(Course course) {
        return courseRepository.save(course);
    }

    public void deleteCourseById(Long courseId) {
        courseRepository.deleteById(courseId);
    }

    @Override
    public long count() {
        return courseRepository.count();
    }
}

File: ch10/Kotlin/course-tracker-kotlin-app/src/main/kotlin/com/manning/sbip/ch10/service/DefaultCourseService.kt
package com.manning.sbip.ch10.service

import com.manning.sbip.ch10.model.Course
import com.manning.sbip.ch10.repository.CourseRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.util.*

@Service
class DefaultCourseService (val courseRepository: CourseRepository) : CourseService {
    override fun createCourse(course: Course): Course {
        return courseRepository.save(course)
    }

    override fun findCourseById(courseId: Long): Optional<Course> {
        return courseRepository.findById(courseId)
    }

    override fun findAllCourses(): Iterable<Course> {
        return courseRepository.findAll()
    }

    override fun updateCourse(course: Course): Course {
        return courseRepository.save(course)
    }

    override fun deleteCourseById(courseId: Long) {
        courseRepository.deleteById(courseId)
    }
}